File: Level-4/quick_sort.c
/*
 * Date: 2017-xx-xx
 *
 * Description:
 * Implement quick sort.
 *
 * Implementation:
 * Uses divide and conquer approach. It divides the array in 2 parts lower part
 * has all elements lesser than all elements at higher part. Partitioning is
 * done across an element called pivot. So after each iteration selected pivot
 * gets placed at correct position in array.
 * In below program last element of sub array is taken as pivot and that is
 * placed at correct place after each call to partition function.
 *
 * Complexity:
 * O(nlogn) time, Best and average case
 * O(n^2) time, Worst case, when array is already sorted in asc or dsc order
 * O(1) space
 */

#include "stdio.h"
#include "stdlib.h"

void swap(int *a, int *b) {
  int t;
  t = *a;
  *a = *b;
  *b = t;
}

/*
 * Partitions the array in 2 parts and places the pivot at correct place in
 * array. Last element is taken as pivot in every call to this function and it
 * is placed at correct position in array.
 *
 * This function has a complexity of O(n).
 *
 * Args:
 * a: Array
 * low: Start index of array
 * high: Last index of array
 *
 * Returns: Partitioning index of array.
 */
int partition(int a[], int low, int high) {
  int i = low;  // Holds counter for smaller values than pivot.
  int j = 0;  // For iterating over array.
  int pivot = a[high];

  for (j = low; j < high; j++) {
    if (a[j] <= pivot) {
      swap(&a[i], &a[j]);
      i++;
    }
  }
  swap(&a[i], &a[high]);
  return i;
}

/*
 * This function does the divide and conquer part. Calls itself recursively with
 * each and every partition of the array.
 *
 * This is the function which governs the complexity of quick sort. For random
 * array this takes O(logn) as array is divided by 2 in each call to this
 * function via pivot from partition function.
 *
 * If input array is sorted (increasing or decreasing) array is not evenly
 * divided, it gets divided with 1 element on left and n - 1 on right side so
 * becomes skewed kind of tree which takes complexity of O(n) instead of
 * O(logn).
 *
 * Args:
 * a: Array
 * low: Start index of array
 * high: End index of array
 */
void quickSort(int a[], int low, int high) {
  int pivot = 0;
  if (low < high) {
    pivot = partition(a, low, high);
    quickSort(a, low, pivot - 1);
    quickSort(a, pivot + 1, high);
  }
}

int main() {
  int i = 0, n = 0;
  int *a = NULL;
  printf("Enter number of elements: ");
  scanf("%d", &n);
  a = (int *)malloc(sizeof(int)*n);
  for (i = 0; i < n; i++) {
    printf("Enter element [%d]: ", i);
    scanf("%d", &a[i]);
  }
  quickSort(a, 0, n - 1);

  printf("Sorted array elements:\n");
  for (i = 0; i < n; i++)
    printf("a[%d] : %d\n", i, a[i]);
  return 0;
}

File: Level-3/quick_sort.py
# Date: 2020-07-19
#
# Description:
# Implement quick sort.
#
# Implementation:
# Uses divide and conquer approach. It divides the array in 2 parts lower part
# has all elements lesser than all elements at higher part. Partitioning is
# done across an element called pivot. So after each iteration selected pivot
# gets placed at correct position in array.
# In below program last element of sub array is taken as pivot and that is
# placed at correct place after each call to partition function.
#
# Complexity:
# O(nlogn) time, Best and average case
# O(n^2) time, Worst case, when array is already sorted in asc or dsc order
# O(1) space

def partion(A, low, high):
    pivot = A[high]
    i = low  # Counts the number of elements less than or equal to pivot
    j = low  # For looping over all elements of A

    while j < high:
        if A[j] <= pivot:
            A[i], A[j] = A[j], A[i]
            i += 1
        j += 1
    A[i], A[high] = A[high], A[i]  # Place pivot element at correct place
    return i

def quick_sort(A, low, high):
    if low < high:
        pivot_idx = partion(A, low, high)
        quick_sort(A, low, pivot_idx - 1)
        quick_sort(A, pivot_idx + 1, high)


def sort(A):
    quick_sort(A, 0, len(A) - 1)
    return A  # To verify assert, otherwise A passed is sorted - inplace sorting

assert sort([2, 3, 10, 5, 6, 1, 3, 9]) == [1, 2, 3, 3, 5, 6, 9, 10]
assert sort([3, 4, 5, 2, 1]) == [1, 2, 3, 4, 5]
assert sort([3, 4, 5, 2, 1, 6]) == [1, 2, 3, 4, 5, 6]
assert sort([]) == []
assert sort([1]) == [1]
assert sort([2, 1]) == [1, 2]
